nano/node/node.cpp

```diff
@@ -3748,6 +3750,89 @@ std::unique_ptr<seq_con_info_component> collect_seq_con_info (active_transaction
 	return composite;
 }
 }
+
+/**
+ * Confirms the height of the block for the block hash passed in. If this has implicitly
+ * confirms other blocks then check if these are open/receive blocks, and if so follow
+ * the source of these blocks iteratively confirm them as well.
+ */
+void nano::node::add_confirmation_heights (nano::transaction const & transaction, nano::block_hash const & hash)
+{
+	std::stack<nano::block_hash, std::vector<nano::block_hash>> open_receive_blocks;
+	auto current = hash;
+
+	do
+	{
+		if (!open_receive_blocks.empty ())
+		{
+			current = open_receive_blocks.top ();
+			open_receive_blocks.pop ();
+
+			auto block (store.block_get (transaction, current));
+			if (block != nullptr)
+			{
+				nano::block_hash source_hash = block->source ();
+				auto source_block (store.block_get (transaction, source_hash));
+				if (source_block != nullptr)
+				{
+					current = source_block->hash ();
+				}
+			}
+		}
+
+		auto hash (current);
+		auto block_height (store.block_account_height (transaction, hash));
+		assert (block_height >= 0);
+		nano::account_info account_info;
+		nano::account account (ledger.account (transaction, hash));
+		release_assert (!store.account_get (transaction, account, account_info));
+		auto confirmation_height = account_info.confirmation_height;
+		if (block_height > confirmation_height)
+		{
+			account_info.confirmation_height = block_height;
+			store.account_put (transaction, account, account_info);
+
+			// Get the difference and check if any of these are recieve blocks
+			auto newly_confirmed_num_blocks = block_height - confirmation_height;
+
+			// Start from the most recent one and work our way through
+			for (uint64_t i = 0; i < newly_confirmed_num_blocks; ++i)
+			{
+				nano::block_sideband sideband;
+				auto block (store.block_get (transaction, current, &sideband));
+				if (block != nullptr)
+				{
+					// First check legacy receive/open
+					auto should_process = (sideband.type == nano::block_type::receive || (sideband.type == nano::block_type::open && !current.is_zero ()));
+					if (!should_process)
```

Comment:
Probably check block type == state before cast?

Reply:
Hm, why check sideband type instead of block type?

Reply:
The `dynamic_pointer_cast`? It does a `dynamic_cast` under the hood and checks for null, so it should be ok to just check for `nullptr` after the cast. I have made it more explicit now though.

Good point, done and that now saves getting the `sideband` object now.